File: backend/agents/insight.py
"""
Insight Agent (Hybrid)
Reads validated structured signals, connects them causally,
and generates an executive intelligence brief.

Uses Gemini API if key provided; falls back to deterministic rule-based engine.
LLM only reasons on VALIDATED signals — never generates raw fake data.
"""
import time
import json
import re
from memory import SharedMemory, ExecutiveSummary, StrategicInsight, Severity, AgentLog

# ── Rule-based fallback ────────────────────────────────────────────────────

def _rule_based(memory: SharedMemory, company: str, industry: str) -> ExecutiveSummary:
    signals = memory.signals
    critical = [s for s in signals if s.severity == Severity.CRITICAL]
    high = [s for s in signals if s.severity == Severity.HIGH]
    opp_signals = [s for s in signals if s.is_opportunity]
    top = signals[0] if signals else None
    top_opp = sorted(opp_signals, key=lambda s: s.score, reverse=True)[0] if opp_signals else None

    insights = []

    if top:
        insights.append(StrategicInsight(
            id="i-1",
            title=f'Critical Signal: "{top.label}"',
            body=(
                f'The "{top.label}" cluster scores {top.score}/100 with {top.volume} posts '
                f'and {top.negativity}% negativity. '
                + (f'Growing at {top.growth_rate}% WoW — actively worsening. '
                   if top.growth_rate > 10 else 'Volume is stable but persistently high. ')
                + (f'Top risk: {top.risks[0]}' if top.risks else '')
            ),
            related_signals=[top.id],
            confidence="high",
            insight_type="risk",
        ))

    if len(critical) + len(high) >= 2:
        cross = (critical + high)[:3]
        insights.append(StrategicInsight(
            id="i-2",
            title="Converging Signals — Systemic Product-Market Gap",
            body=(
                f'{len(cross)} high-severity clusters converging simultaneously: '
                + ", ".join(f'"{s.label}"' for s in cross)
                + f'. Covering {sum(s.volume for s in cross)} mentions, this pattern indicates '
                  'a fundamental gap between product promises and delivered experience.'
            ),
            related_signals=[s.id for s in cross],
            confidence="high",
            insight_type="trend",
        ))

    if top_opp:
        insights.append(StrategicInsight(
            id="i-3",
            title=f'Highest-ROI Opportunity: "{top_opp.label}"',
            body=(
                f'{top_opp.volume} users are actively expressing pain around "{top_opp.label}". '
                + (f'Recommended action: {top_opp.opportunities[0]}' if top_opp.opportunities else '')
                + ' This is a validated market gap with measurable user demand.'
            ),
            related_signals=[top_opp.id],
            confidence="medium",
            insight_type="opportunity",
        ))

    top_risk_signal = critical[0] if critical else (high[0] if high else top)
    return ExecutiveSummary(
        headline=(
            f'{company}: {len(critical)} critical + {len(high)} high-severity signals '
            f'across {len(signals)} validated topics in {industry}'
        ) if signals else f'{company}: Insufficient signal data for analysis',
        top_opportunity=top_opp.opportunities[0] if (top_opp and top_opp.opportunities) else "Monitor community sentiment for emerging product gaps",
        top_risk=top_risk_signal.risks[0] if (top_risk_signal and top_risk_signal.risks) else "No critical risks detected",
        insights=insights,
        generated_at=time.time(),
        is_ai_generated=False,
    )


# ── Gemini-backed generation ───────────────────────────────────────────────

def _gemini(memory: SharedMemory, company: str, industry: str, api_key: str) -> ExecutiveSummary:
    import google.generativeai as genai

    genai.configure(api_key=api_key)
    model = genai.GenerativeModel("gemini-1.5-flash")

    top5 = memory.signals[:5]
    signal_summary = "\n".join(
        f'- "{s.label}" | Score: {s.score}/100 | {s.severity.value.upper()} | '
        f'Vol: {s.volume} | Neg: {s.negativity}% | Growth: {s.growth_rate}%\n'
        f'  Risks: {"; ".join(s.risks[:2])}\n'
        f'  Opportunities: {"; ".join(s.opportunities[:2])}'
        for s in top5
    )

    prompt = f"""You are a senior R&D intelligence analyst. Analyze these validated market signals for {company} ({industry}) and generate a JSON executive brief.

VALIDATED SIGNALS (from real Reddit/NewsAPI data, TF-IDF clustered, sentiment-scored):
{signal_summary}

Return ONLY valid JSON in this exact structure:
{{
  "headline": "One specific, compelling sentence about {company}'s intelligence picture",
  "top_risk": "The single most urgent risk with specific action",
  "top_opportunity": "The single highest-ROI product opportunity",
  "insights": [
    {{
      "id": "i-1",
      "title": "Specific title",
      "body": "2-3 sentences with causal reasoning based only on the data above",
      "related_signals": ["signal-0"],
      "confidence": "high",
      "type": "risk"
    }},
    {{...}},
    {{...}}
  ]
}}

Generate exactly 3 insights. Base everything ONLY on the validated signals above. No hallucination."""

    response = model.generate_content(prompt)
    raw = response.text.strip()

    # Extract JSON block
    match = re.search(r"\{.*\}", raw, re.DOTALL)
    if not match:
        raise ValueError("Gemini returned no valid JSON")

    parsed = json.loads(match.group())

    insights = [
        StrategicInsight(
            id=ins.get("id", f"i-{i}"),
            title=ins.get("title", ""),
            body=ins.get("body", ""),
            related_signals=ins.get("related_signals", []),
            confidence=ins.get("confidence", "medium"),
            insight_type=ins.get("type", "trend"),
        )
        for i, ins in enumerate(parsed.get("insights", []))
    ]

    return ExecutiveSummary(
        headline=parsed.get("headline", ""),
        top_opportunity=parsed.get("top_opportunity", ""),
        top_risk=parsed.get("top_risk", ""),
        insights=insights,
        generated_at=time.time(),
        is_ai_generated=True,
    )


# ── Public interface ───────────────────────────────────────────────────────

def run_insight(memory: SharedMemory, company: str, industry: str, gemini_key: str = "") -> None:
    start = time.time()
    memory.agent_logs.append(AgentLog("Insight Agent", "running", "Synthesizing signals into executive intelligence...", time.time()))

    if gemini_key and memory.signals:
        try:
            memory.agent_logs.append(AgentLog("Insight Agent", "running", "Calling Gemini 1.5 Flash for strategic analysis...", time.time()))
            memory.executive_summary = _gemini(memory, company, industry, gemini_key)
            duration = int((time.time() - start) * 1000)
            memory.agent_logs.append(AgentLog("Insight Agent", "done", "AI executive summary generated ✓", time.time(), duration))
            return
        except Exception as e:
            print(f"[InsightAgent] Gemini failed: {e}, using rule-based fallback")
            memory.agent_logs.append(AgentLog("Insight Agent", "running", f"Gemini failed ({e}), using deterministic fallback...", time.time()))

    memory.executive_summary = _rule_based(memory, company, industry)
    duration = int((time.time() - start) * 1000)
    mode = "AI" if (memory.executive_summary.is_ai_generated) else "rule-based"
    memory.agent_logs.append(AgentLog("Insight Agent", "done", f"Executive summary generated ({mode}) ✓", time.time(), duration))

File: Insight/src/agents/insight.ts
/**
 * Insight Agent (Tier 3 — Hybrid AI)
 * Reads structured memory, connects related signals, produces strategic insights.
 * 
 * If LLM is enabled (OpenAI/Anthropic API key provided), calls the API.
 * Otherwise falls back to deterministic rule-based insight templates.
 * 
 * This mirrors World Monitor's server/_shared/llm.ts pattern.
 */
import type { Signal, ExecutiveSummary, StrategicInsight, WorkspaceConfig } from '@/types';

// ── Rule-based fallback insight generator ─────────────────────────────────

function buildRuleBasedInsights(signals: Signal[], config: WorkspaceConfig): ExecutiveSummary {
  const critical = signals.filter(s => s.severity === 'critical');
  const high = signals.filter(s => s.severity === 'high');
  const opportunities = signals.filter(s => s.isOpportunity);
  const topSignal = signals[0];
  const topOpp = opportunities.sort((a, b) => b.score - a.score)[0];

  const insightTemplates: StrategicInsight[] = [];

  // Insight 1: Top pain cluster
  if (topSignal) {
    insightTemplates.push({
      id: 'insight-1',
      title: `Critical UX Pain: "${topSignal.label}"`,
      body: `The "${topSignal.label}" cluster has emerged as the highest-priority signal with a score of ${topSignal.score}/100. ` +
        `With ${topSignal.volume} posts and ${topSignal.negativity}% negativity rate, this represents a validated systemic pain point ` +
        `in your ${config.industry} target market. ` +
        (topSignal.growthRate > 0
          ? `The ${topSignal.growthRate}% week-over-week growth rate indicates this pain is actively worsening — prioritize immediately.`
          : `This pain appears stable but persistent — a product investment here would provide immediate differentiation.`),
      relatedSignals: [topSignal.id],
      confidence: 'high',
      type: 'risk',
    });
  }

  // Insight 2: Cross-cluster pattern
  if (critical.length + high.length >= 2) {
    const crossSignals = [...critical, ...high].slice(0, 3);
    const crossLabels = crossSignals.map(s => `"${s.label}"`).join(', ');
    insightTemplates.push({
      id: 'insight-2',
      title: 'Converging Pain Signals Indicate Product-Market Misalignment',
      body: `Multiple high-severity signals (${crossLabels}) share a common root: ` +
        `users are not getting the value they were promised at the point of sale. ` +
        `The pattern across ${crossSignals.reduce((s, c) => s + c.volume, 0)} posts suggests ` +
        `a fundamental gap between marketing messaging and product experience. ` +
        `Consider auditing the entire post-signup journey against user expectations.`,
      relatedSignals: crossSignals.map(s => s.id),
      confidence: 'high',
      type: 'trend',
    });
  }

  // Insight 3: Opportunity signal
  if (topOpp) {
    insightTemplates.push({
      id: 'insight-3',
      title: `Highest-ROI Opportunity: "${topOpp.label}"`,
      body: `Among all validated signals, "${topOpp.label}" presents the strongest product opportunity. ` +
        `With ${topOpp.volume} vocal users expressing this pain and ${topOpp.opportunities.length} identified product gaps, ` +
        `a targeted feature investment here could convert frustrated users into product advocates. ` +
        `Recommended next step: conduct 5 customer discovery interviews focused specifically on this theme.`,
      relatedSignals: [topOpp.id],
      confidence: 'medium',
      type: 'opportunity',
    });
  }

  const topRisk = critical[0] ?? high[0] ?? signals[0];
  const headline = topSignal
    ? `${config.companyName || 'Your Market'}: ${critical.length} critical signals detected across ${signals.length} validated topics`
    : 'Analysis complete — insufficient data for strong signals';

  return {
    headline,
    topOpportunity: topOpp?.opportunities[0] ?? 'Invest in user onboarding experience based on friction signals',
    topRisk: topRisk?.risks[0] ?? 'Monitor growing complaint velocity before it impacts retention',
    insights: insightTemplates,
    generatedAt: Date.now(),
    isAiGenerated: false,
  };
}

// ── LLM-backed insight generator ──────────────────────────────────────────

async function generateLLMInsights(
  signals: Signal[],
  config: WorkspaceConfig,
): Promise<ExecutiveSummary> {
  const top5 = signals.slice(0, 5);
  const signalSummary = top5.map(s =>
    `- Signal "${s.label}" (Score: ${s.score}/100, Severity: ${s.severity}, Volume: ${s.volume} posts, Negativity: ${s.negativity}%, Growth: ${s.growthRate}%)\n  Risks: ${s.risks.slice(0, 2).join('; ')}\n  Opportunities: ${s.opportunities.slice(0, 2).join('; ')}`
  ).join('\n');

  const prompt = `You are an expert R&D intelligence analyst. Analyze the following market signals for a ${config.industry} company named ${config.companyName} and generate a concise executive intelligence brief.

TOP SIGNALS:
${signalSummary}

Generate a JSON response with this exact structure:
{
  "headline": "One compelling sentence summarizing the overall intelligence picture",
  "topOpportunity": "The single highest-ROI product opportunity",
  "topRisk": "The single most critical risk to address",
  "insights": [
    {
      "id": "insight-1",
      "title": "Insight title",
      "body": "2-3 sentence strategic analysis with causal reasoning",
      "relatedSignals": ["signal id"],
      "confidence": "high|medium|low",
      "type": "risk|opportunity|trend"
    }
  ]
}
Generate exactly 3 insights. Be specific, actionable, and connect signals causally.`;

  const apiUrl = `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=${config.llmApiKey}`;

  const body = JSON.stringify({
    contents: [{ parts: [{ text: prompt }] }],
    generationConfig: {
      responseMimeType: "application/json",
    }
  });

  const res = await fetch(apiUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body
  });

  if (!res.ok) {
    const errText = await res.text();
    throw new Error(`Gemini API error ${res.status}: ${errText}`);
  }

  const data = await res.json();
  const raw = data.candidates?.[0]?.content?.parts?.[0]?.text ?? '{}';

  const parsed = JSON.parse(raw) as Partial<ExecutiveSummary>;
  return {
    headline: parsed.headline ?? '',
    topOpportunity: parsed.topOpportunity ?? '',
    topRisk: parsed.topRisk ?? '',
    insights: (parsed.insights ?? []).map((ins, i) => ({
      id: `insight-${i}`,
      title: (ins as StrategicInsight).title ?? '',
      body: (ins as StrategicInsight).body ?? '',
      relatedSignals: (ins as StrategicInsight).relatedSignals ?? [],
      confidence: (ins as StrategicInsight).confidence ?? 'medium',
      type: (ins as StrategicInsight).type ?? 'trend',
    })),
    generatedAt: Date.now(),
    isAiGenerated: true,
  };
}

// ── Public API ─────────────────────────────────────────────────────────────

export async function runInsightAgent(
  signals: Signal[],
  config: WorkspaceConfig,
  onProgress: (msg: string) => void,
): Promise<ExecutiveSummary> {
  onProgress('Insight Agent: Connecting signals and generating strategic intelligence...');
  await new Promise(r => setTimeout(r, 300));

  if (config.llmApiKey) {
    try {
      onProgress('Insight Agent: Calling LLM for strategic analysis...');
      const summary = await generateLLMInsights(signals, config);
      onProgress('Insight Agent: AI executive summary generated ✓');
      return summary;
    } catch (err) {
      console.warn('[InsightAgent] LLM failed, falling back to rule-based:', err);
      onProgress('Insight Agent: LLM failed — using rule-based fallback...');
    }
  }

  const summary = buildRuleBasedInsights(signals, config);
  onProgress('Insight Agent: Strategic insights generated ✓');
  return summary;
}
